packages/dht/src/dht/discovery/DiscoverySession.ts

```diff
@@ -112,17 +105,13 @@ export class DiscoverySession {
         if (this.config.peerManager.getContactCount(this.config.contactedPeers) === 0) {
             return
         }
-        // TODO add abortController and signal it in stop()
-        await runAndWaitForEvents3<DiscoverySessionEvents>(
-            [this.findMoreContacts.bind(this)],
-            [[this.emitter, 'discoveryCompleted']],
-            timeout
-        )
-    }
 
-    public stop(): void {
-        this.stopped = true
-        this.emitter.emit('discoveryCompleted')
-        this.emitter.removeAllListeners()
+        setImmediate(() => {
+            this.findMoreContacts()
+        })
+        await Promise.race([
```

Comment:
Would something this work here?

```suggestion
        await withTimeout(this.doneGate.waitUntilOpen(), timeout, 'discovery session timed out', this.config.abortSignal)
```

Reply:
Good improvement 👍 Fixed